lightning/src/routing/router.rs

```diff
@@ -994,42 +1000,87 @@ impl cmp::PartialOrd for RouteGraphNode {
 ///
 /// Used to construct a [`PathBuildingHop`] and to estimate [`EffectiveCapacity`].
 #[derive(Clone, Debug)]
-enum CandidateRouteHop<'a> {
+pub enum CandidateRouteHop<'a> {
 	/// A hop from the payer, where the outbound liquidity is known.
 	FirstHop {
+		/// The channel details of the first hop.
+		/// `get_outbound_payment_scid` is assumed to always return Some(scid) - this assumption is checked in
+		/// `find_route` method.
 		details: &'a ChannelDetails,
+		/// The node id of the payer.
+		///
+		/// Can be accessed via `source` method.
+		node_id: NodeId
 	},
 	/// A hop found in the [`ReadOnlyNetworkGraph`], where the channel capacity may be unknown.
 	PublicHop {
+		/// The channel info of the hop.
 		info: DirectedChannelInfo<'a>,
+		/// The short_channel_id of the channel.
 		short_channel_id: u64,
+		/// The node id leading to the current hop in route.
+		source_node_id: NodeId,
+		/// The node id of current hop in route.
+		target_node_id: NodeId,
 	},
 	/// A hop to the payee found in the BOLT 11 payment invoice, though not necessarily a direct
 	/// channel.
 	PrivateHop {
+		/// Hint provides information about a private hop, needed while routing through a private
+		/// channel.
 		hint: &'a RouteHintHop,
+		/// The node id of current hop in route.
+		target_node_id: NodeId,
 	},
 	/// The payee's identity is concealed behind blinded paths provided in a BOLT 12 invoice.
 	Blinded {
+		/// Hint provides information about a blinded hop, needed while routing through a blinded
+		/// channel.
+		/// `BlindedPayInfo` provides information needed about the
+		/// payment while routing through a blinded
+		/// path.
+		/// `BlindedPath` is the blinded path to the destination.
 		hint: &'a (BlindedPayInfo, BlindedPath),
+		/// The index of the hint in the original list of blinded hints.
+		/// Provided to uniquely identify a hop as we are
+		/// route building.
 		hint_idx: usize,
+		/// The node id of current hop in route.
+		///
+		/// This will always equal to `maybe_dummy_payee_node_id` in `find_route` method.
+		target_node_id: NodeId,
```

Comment:
ok few questions here because this thing is a kinda big.
`add_entry!` macro have also `target` and `source` inputs as well as `Candidate` - after adding `fn target` and `fn source` to `Candidate`, we should remove the source and target from `add_entry!` correct?
regarding `maybe_dummy_payee_node_id`,  should I always run `candidate.target().unwrap_or(maybe_dummy_payee_node_id)` ?
we have this bit 
`if ordered_hops.last().unwrap().0.candidate.target().unwrap() == maybe_dummy_payee_node_id`
  do we wanna keep this? or do `if ....candidate.target().is_none()....` ?
  
last question (: 
the `id` function takes a direction as an input which is `sourc < target` should I keep that input as is or make it optional? because we have some places in the code were we call it with `source < target` but other places where we just give `true/false`..

Reply:
pushed some code with the stuff I mentioned in prev comment^.. please let me know what u think

Reply:
the question regarding the `id` function is because we dont need to provide `source < target` now because `Candidate` has this info (see code please)

Reply:
> ok few questions here because this thing is a kinda big. `add_entry!` macro have also `target` and `source` inputs as well as `Candidate` - after adding `fn target` and `fn source` to `Candidate`, we should remove the source and target from `add_entry!` correct?

Ideally, yes, though there is the blinded case that needs to be considered, as you mention below.

> regarding `maybe_dummy_payee_node_id`, should I always run `candidate.target().unwrap_or(maybe_dummy_payee_node_id)` ? we have this bit `if ordered_hops.last().unwrap().0.candidate.target().unwrap() == maybe_dummy_payee_node_id` we cant keep it like this as its not correct, should I do `if ....candidate.target().is_none()....` ?

`maybe_dummy_payee_node_id` always seemed a little hacky to me. @TheBlueMatt Is there any reason we can't check against `None` instead?

> 
> last question (: the `id` function takes a direction as an input which is `sourc < target` should I keep that input as is or make it optional? because we have some places in the code were we call it with `source < target` but other places where we just give `true/false`..

Huh, it's kinda error-prone that it takes a direction in the first place given the `source` and `target` relationship is what defines the direction (i.e., a candidate's direction, and thus id, should never change). So someone could mix it up if they need to pass it in. Seems like the direction could be derived from the candidate now, though.

Looks like we pass in `true` or `false` when we need a `CandidateHopId` in both directions, we we could add a method to `CandidateHopId` gives the `CandidateHopId` for the opposite direction. That way, we should never need to pass in a direction to obtain a candidate's id.

Also, we use `true` unilaterally in some places, which seems wrong? Or maybe it just doesn't matter at that point and we need some id, if not the most accurate one?

Reply:
I see... yeah, hard to say what's best. Having an internal `target` function returning `NodeId` instead of `Option<NodeId>` may be better so it's all defined in one place then (opposed to `unwrap_or(maybe_dummy_payee_node_id)`). But there are places where we expect the candidate to not be blinded, so it's questionable if we should be replacing those with `maybe_dummy_payee_node_id`. I left some comments where this may be the case or where `maybe_dummy_payee_node_id` isn't strictly necessary. @jbesraa  Let's see what remains after addressing those comments.